File: backend/src/guitar_player/schemas/job.py
"""Job request/response schemas."""

import uuid
from datetime import datetime

from pydantic import BaseModel


class CreateJobRequest(BaseModel):
    song_id: uuid.UUID
    descriptions: list[str]
    mode: str = "isolate"


class JobResultEntry(BaseModel):
    description: str
    target_key: str | None = None
    residual_key: str | None = None


class ActiveJobInfo(BaseModel):
    """Lightweight summary of an active job, embedded in SongDetailResponse."""

    id: uuid.UUID
    status: str
    progress: int = 0
    stage: str | None = None


class JobResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    song_id: uuid.UUID
    status: str
    progress: int | None = None
    stage: str | None = None
    descriptions: list[str] | None = None
    mode: str | None = None
    error_message: str | None = None
    results: list[JobResultEntry] | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}

File: frontend/src/types/job.ts
/** Const object for job status values. Use `JobStatus.PENDING` etc. instead of raw strings. */
export const JobStatus = {
  PENDING: 'PENDING',
  PROCESSING: 'PROCESSING',
  COMPLETED: 'COMPLETED',
  FAILED: 'FAILED',
} as const

export type JobStatus = (typeof JobStatus)[keyof typeof JobStatus]

export interface JobResult {
  description: string
  target_key: string
  residual_key: string
}

export interface Job {
  id: string
  user_id: string
  song_id: string
  status: JobStatus
  progress?: number | null
  stage?: string | null
  descriptions: string[]
  mode: string
  error_message: string | null
  results: JobResult[] | null
  created_at: string
  updated_at: string
  completed_at: string | null
}
